api/models/metadata.go

```diff
@@ -0,0 +1,230 @@
+package models
+
+import (
+	"encoding/json"
+	"fmt"
+	"bytes"
+	"database/sql/driver"
+	"errors"
+)
+
+// Metadata encapsulates key-value metadata. The structure is immutable via its public API and nil-safe for its contract
+// permissive nilability is here to simplify updates and reduce the need for nil handling in extensions - metadata should be updated by over-writing the original object:
+//  oldMd := EmptyMetadata()
+//  newMd := oldMd.With("fooKey",1)
+//  // old MD remains empty
+// Metadata is lenable
+type Metadata map[string]*metadataValue
+
+// metadataValue encapsulates a metadata value in the metadata map,
+// This is stored in its un-parsed JSON format for later (re-) parsing into specific structs or values
+// metadataValue objects are  immutable after JSON load
+type metadataValue struct {
+	val []byte
+}
+
+const (
+	maxMetadataValueBytes = 512
+	maxMetadataKeyBytes   = 128
+	maxMetadataKeys       = 100
+)
+
+// Equals is defined based on un-ordered k/v comparison at of the metadata keys and  (compacted) values of metadata, JSON object-value equality for values is property-order dependent
+func (m Metadata) Equals(other Metadata) bool {
+	if len(m) != len(other) {
+		return false
+	}
+	for k1, v1 := range m {
+		v2, _ := other[k1]
+		if !bytes.Equal(v1.val, v2.val) {
+			return false
+		}
+	}
+	return true
+}
+
+func EmptyMetadata() Metadata {
+	return nil
+}
+
+func (mv *metadataValue) String() string {
+	return string(mv.val)
+}
+
+func (v *metadataValue) MarshalJSON() ([]byte, error) {
+	return []byte(v.val), nil
+}
+
+func (v *metadataValue) isEmptyValue() bool {
+	return string(v.val) == "\"\""
+}
+
+// UnmarshalJSON compacts metadata values but does not alter key-ordering for keys
+func (v *metadataValue) UnmarshalJSON(val []byte) (error) {
+	buf := bytes.Buffer{}
+	err := json.Compact(&buf, val)
+	if err != nil {
+		return err
+	}
+	if err != nil {
+		return err
+	}
+	v.val = buf.Bytes()
+	return nil
+}
+
+func validateField(key string, value *metadataValue) APIError {
+
```

Comment:
Should non-ascii characters/non printable chars/whitespace be allowed?

Reply:
yeah , good point - it's probabably safer to restrict than to allow here. 

I originally supposed that keys were "any utf8 string up to 128 bytes" (#789 )  -